### PR TITLE
Feature/qti 168 expand gi archive on upload

### PR DESCRIPTION
**Pull request by wtritch**
This PR adds an additional API endpoint to handle uploading GI archives.
On upload the GI archive is cracked open, parsed, and validated.

The uploaded `.gax` is embedded within the item model.
The gax-referenced resources are uploaded independently added as Attachment instances to the item-content.

The goals are:
1. The item.json contains everything needed to create a SAAIF/QTI assessment item XML document
2. All independent resources (images, audio files, etc) can be independently requested from S3

GI Archive expansion PR set:
AP_ItemManagementService: https://github.com/SmarterApp/AP_ItemManagementService/pull/189
AP_Common changes: https://github.com/SmarterApp/AP_Common/pull/558
AP_ItemAuthoringTool changes: https://github.com/SmarterApp/AP_ItemAuthoringTool/pull/786

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -12,7 +12,7 @@ group=org.opentestsystem.ap
 
 version=0.5.50-SNAPSHOT
 
-commonVersion=0.5.64
+commonVersion=0.5.65-SNAPSHOT
```
Comment (wtritch) [inline gradle.properties]: I'll update this after releasing AP_Common

File: src/main/java/org/opentestsystem/ap/ims/rest/ItemManagerApi.java [hunk 2]
```diff
@@ -354,6 +356,16 @@ public String readFileContents(@ModelAttribute final UploadFile uploadFile) {
...
 
+    /**
+     * @param uploadFile The uploaded file and associated data.
+     */
+    @PostMapping("/{itemId}/gi/archive")
+    public GiItemContent attachGiArchive(@ModelAttribute final UploadFile uploadFile) {
+        log.debug("Uploading file");
+        uploadFile.setSection(SECTION_CORE);
+        return attachmentService.saveGiArchive(uploadFile);
+    }
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/ims/rest/ItemManagerApi.java]: new simple API for uploading a GI archive

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java [hunk 3]
```diff
@@ -16,43 +16,196 @@
...
 
+    /**
+     * Validates, processes, and uploads the given GI ZIP archive
+     * containing a .GAX grid item definition XML and required
+     * resources (images.)
+     *
+     * @param archiveUpload The uploaded GI archive
+     * @return  The updated GiItemContent referencing the uploaded GI archive resource(s)
+     */
+    public GiItemContent saveGiArchive(final UploadFile archiveUpload) {
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: This method is the meat of the PR set.

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java [hunk 4]
```diff
@@ -16,43 +16,196 @@
...
+                            archiveUpload.getItemId(),
+                            userTransaction.getSection(),
+                            resource.getFileName().toString(),
+                            resource
+                    ))
+                    .collect(Collectors.toList());
+
+            // Include translated GAX resource overrides
+            // TODO These should probably be represented by the UI and model as translation resource overrides
+            //  rather than reply on zip-contents file-naming conventions.
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: Any GI resource can have a localized override version by including an additional resource in the zip appended with "_ESN"
For example:
Original: `my-background-image.png`
Spanish override: `my-background-image_ESN.png`

File: src/test/java/org/opentestsystem/ap/ims/service/AttachmentServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        final ItemTransaction transaction = ItemTransaction.builder()
+                .commitId(UUID.randomUUID())
+                .section(SECTION_CORE)
+                .userName(ITEM_BANK_USER.getUserName())
+                .build();
+        when(itemManager.findUserTransaction(ITEM_BANK_USER, ItemId)).thenReturn(transaction);
+    }
+
+    @Test
+    public void itShouldThrowIfUserNotEditing() throws Exception {
```
Comment (wtritch) [inline src/test/java/org/opentestsystem/ap/ims/service/AttachmentServiceTest.java]: These tests describe the functional requirements for uploading a GI archive.

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java [hunk 6]
```diff
@@ -16,43 +16,196 @@
...
+     */
+    public GiItemContent saveGiArchive(final UploadFile archiveUpload) {
+        archiveUpload.validate();
+
+        final ItemBankUser user = securityUtil.getItemBankUser();
+        final GiItem item = (GiItem) itemManager.findItem(user, archiveUpload.getItemId()).getItem();
+        final GiItemContent contentModel = item.getCore().getEn();
+
+        final ItemTransaction userTransaction = itemManager.findUserTransaction(user, archiveUpload.getItemId());
+        if (Objects.isNull(userTransaction)) {
```
Comment (esotrope) [inline src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: what does this util do beyond `== null` ?

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java [hunk 7]
```diff
@@ -16,43 +16,196 @@
...
+
+            final List<Attachment> translationResources = uploadedResources.entrySet().stream()
+                    .filter(entry -> {
+                        final String uploadedFilename = removeExtension(entry.getKey());
+                        final Matcher matcher = TRANSLATED_RESOURCE_FILENAME.matcher(uploadedFilename);
+                        if (!matcher.matches()) {
+                            return false;
+                        }
+
+                        final String referencedFilename = matcher.group(1);
```
Comment (esotrope) [inline src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: nit: separate method might be good for this

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java [hunk 8]
```diff
@@ -16,43 +16,196 @@
...
+                        return referencedFilenameBases.contains(referencedFilename);
+                    })
+                    .map(entry -> uploadCoreFile(
```
Comment (esotrope) [inline src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: side effects should never occur within a map call, can you create the objects in a map call and then for-each that for the ops?

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java (hunk 6 above)
Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: Nothing, but in this project it seems to be the preferred way to null-check.  When in Rome...

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java (hunk 8 above)
Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: I could, but then my forEach needs to append the response from `uploadCoreFile(...)` to an instantiated list. ...which is basically a map with an additional data structure... thoughts?

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java (hunk 7 above)
Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: Yeah... breaking some of this out into sub-methods might help clarity.

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java (hunk 8 above)
Comment (esotrope) [reply src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: Hm. maybe just add a comment here to that affect for now

Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: Maybe it'll look cleaner after I break some of this out into submethods

Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: I'm still uploading in a `.map` after the reorg but it's the only thing happening in the map, so hopefully it reads a little cleaner.

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java [hunk 9]
```diff
@@ -295,4 +263,46 @@ private Attachment uploadCoreFile(final String itemId, final String section, fin
...
+                    if (!matcher.matches()) {
+                        return false;
+                    }
+
+                    final String referencedFilename = matcher.group(1);
+                    return referencedFilenameBases.contains(referencedFilename);
+                })
+                .map(Map.Entry::getValue)
+                .collect(Collectors.toList());
+    }
```
Comment (esotrope) [inline src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: 👍

File: src/main/java/org/opentestsystem/ap/ims/rest/ItemManagerApi.java [hunk 10]
```diff
@@ -354,6 +356,16 @@ public String readFileContents(@ModelAttribute final UploadFile uploadFile) {
     // Attach files to an item
     // ------------------------------------------------------------------------
 
+    /**
+     * @param uploadFile The uploaded file and associated data.
+     */
+    @PostMapping("/{itemId}/gi/archive")
+    public GiItemContent attachGiArchive(@ModelAttribute final UploadFile uploadFile) {
+        log.debug("Uploading file");
```
Comment (agorina) [inline src/main/java/org/opentestsystem/ap/ims/rest/ItemManagerApi.java]: Nit: maybe add a file name to the log?

File: src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java [hunk 11]
```diff
@@ -132,4 +253,56 @@ public void processTestAdministrationData(final UploadFile uploadFile) {
         log.debug("Uploading file to bulk upload test administration data ");
         this.bulkUpdateService.processTestAdministrationData(securityUtil.getItemBankUser(), uploadFile.getFile());
     }
+
+    private Attachment uploadCoreFile(final String itemId, final String section, final String filename, final Path file) {
+        attachmentValidator.validateFileNameDoesNotHaveIllegalCharacters(filename);
```
Comment (agorina) [inline src/main/java/org/opentestsystem/ap/ims/service/AttachmentService.java]: I am impressed by the length of this method name!

File: src/main/java/org/opentestsystem/ap/ims/rest/ItemManagerApi.java (hunk 10 above)
Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/ims/rest/ItemManagerApi.java]: good call, done

**Pull request merged**